distributed/tests/test_worker.py

```diff
@@ -1715,6 +1714,68 @@ def bad_heartbeat_worker(*args, **kwargs):
+                # error tolerance on estimated duration (needed cause exec times
+                # depend on worker heartbeats)
+                tolerance = 0.2
+                ws = s.workers[w.address]
+                # initialize a task to set initial duration average to 0.1
+                r = c.submit(slowinc, 2.2, delay=0.1)
+                await r
+                # sleep times for some tasks with a shifting average duration
+                times = [5.]*3 + [11.]*3 + [15.]
+                # Store expected task duration after t seconds 
+                # (NOTE t is the _sum_ of sleep times).
+                # at any t:
+                #   - if t < 2 * cur_avg then duration = cur_avg
+                #   - if t > 2 * cur_avg then duration = 2 * t
+                # the cur_avg will depend on tasks that already completed by t
+                sleep2duration = [
+                        # avg at this point will be 0.1
+                        (0.13, 0.1), 
+                        # we're in the outlier regime
+                        (1, 2), # 1s
+                        (1, 4), # 2s
+                        (2, 8), # 4s
+                        # the new average is ~4.5 so no longer an outlier
+                        (3, 5.), # 7s
+                        # we're an outlier again so expect 2x time
+                        (3.5, 21), # 10.5s
+                        # back to average
+                        (1.5, 11.), # 12s
+                        (3, 11.), # 15s
+                        ]
+                key2time=dict()
+                # start a bunch of tasks with different sleep intervals
+                for n,td in enumerate(times):
+                    r = c.submit(slowinc, n, delay=td)
+                    key2time[r.key] = (td,r)
+                t0=time()
+                # every time we wait up we check that any running tasks have
+                # apporpriate durations
+                for (sleep_time, exp_avg) in sleep2duration:
+                    await asyncio.sleep(sleep_time)
+                    # transition any finished tasks so their averages get
+                    # updated in the scheduler
+                    for key,(td,r) in key2time.items():
+                        if td < sleep_time:
+                            await r
+                    for ts in ws.processing:
+                        #logger.info(f"***************************************************************")
+                        #logger.info(f"TIMES ***  {time() - t0}  t={sleep_time}, avg[{ts}] = {ws.processing[ts]}")
+                        #logger.info(f"***************************************************************")
+                        expected_duration = ws.processing[ts]
+                        # expected duraction should be close to d0
+                        assert expected_duration > (1-tolerance) * exp_avg
+                        assert expected_duration < (1+tolerance) * exp_avg
```

Comment:
I think that we can test this more simply, and with less reference to internal API (most Dask distributed tests try very hard not to reference the internal API, which gives us freedom to change that API smoothly in the future).  In general we want to test behaviors, not state.  In this case the behavior that we want to test is, I think, that Dask does load balancing even when the estimate is very wrong.

Here is maybe a short attempt at a test.

```python
@gen_cluster(client=True, nthreads=[("127.0.0.1", 1)] * 2)
def test_balance_with_longer_task(c, s, a, b):
    await client.submit(slowinc, 0, delay=0)  # scheduler learns that slowinc is very fast

    x = client.scatter(np.arange, 100000, workers=[a.address])  # put some weight on a
    y = client.submit(slowinc, 1, delay=5, workers=[a.address])  # a surprisingly long task
    z = client.submit(inc, x, workers=[a.address], allow_other_workers=True)  # a task after y, suggesting a, but open to b

    await z
    assert z.key in b.data
```

I *think* that this will fail today (it would be good to be sure)

Reply:
@selshowk and I were able to talk about this test offline on Friday (summarizing the output of that conversation here). 

There are a few different [types of priorities that are used to determine the order in which tasks are executed](https://distributed.dask.org/en/latest/priority.html) (e.g. user-provided priority, priority based on graph structure, etc.). For tasks which are submitted rapidly, and all other priorities are the same, we use the order in which tasks arrive at a worker as a tie breaker. This tie breaking is implemented using a last-in-first-out policy.

So in the above test, `z` will actually start running before `y` (due to LIFO tie breaking). To ensure `y` starts running before `z`, we can specify our own user-provided priority: 

```python
@gen_cluster(client=True)
async def test_balance_with_longer_task(c, s, a, b):
    np = pytest.importorskip("numpy")

    await c.submit(slowinc, 0, delay=0)  # scheduler learns that slowinc is very fast
    x = await c.scatter(np.arange(10000), workers=[a.address])
    y = c.submit(
        slowinc, 1, delay=5, workers=[a.address], priority=1
    )  # a surprisingly long task
    z = c.submit(
        inc, x, workers=[a.address], allow_other_workers=True, priority=0
    )  # a task after y, suggesting a, but open to b
    await z
    assert z.key in b.data
```

> I think that this will fail today (it would be good to be sure)

I can confirm this fails on the main branch, but passes here